### PR TITLE
Enable min & max for Float16 & BFloat16

### LINKED ISSUE #50790: Implement torch.max for float16 and bfloat16 on CPU and CUDA
## 🚀 Feature

Add support for `torch.max` with:
* CUDA bfloat16
* CPU float16 and bfloat16

## Motivation

Currently, `torch.max` has support for CUDA float16:
```
>>> torch.rand(10, dtype=torch.float16, device='cuda').max()
tensor(0.8530, device='cuda:0', dtype=torch.float16)
```

But all three other combinations of CPU/CUDA and float16/bfloat16 are not supported:
```
>>> torch.rand(10, dtype=torch.float16).max()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: "max_all" not implemented for 'Half'
>>> torch.rand(10, dtype=torch.bfloat16).max()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: "max_all" not implemented for 'BFloat16'
>>> torch.rand(10, dtype=torch.bfloat16, device='cuda').max()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: "max_all_cuda" not implemented for 'BFloat16'
```

### PR DESCRIPTION
**Pull request by imaginary-person**
Fixes #50790.

### SUMMARY
Added `min()` & `max()` support for `Float16` & `BFloat16`.
CUDA already supported these ops on `Float16`, so the other three combinations had to be enabled.
`OpInfo`s for `min` & `max` were also added, and their sample inputs were removed from `method_tests()`.

### MORE INFO
The (slightly) long-term goal is to add dispatch for `min()` & `max()` related operations on CPU & CUDA for `Float16` & `BFloat16`,
wherever they aren't present already:
1. `amin()` 
2. `argmax()` 
3. `amax()` 
4. `argmin()`
5. `torch._aminmax()`
6. `torch.clamp()` on CPU. Was already supported on CUDA
7. `min()` (in this PR)
8. `max()` (in this PR)
9. `minimum()` - already enabled, but OpInfo can be added
10. `maximum()` - already enabled, but OpInfo can be added
11. `fmax()` - already enabled, but OpInfo can be added
12. `fmin()` - already enabled, but OpInfo can be added

I'll submit separate PRs for the other ops.

### CODE REVIEW

File: aten/src/ATen/native/cuda/ReduceMinMaxKernel.cu [hunk 1]
```diff
@@ -74,6 +74,10 @@ void argmax_kernel_cuda(TensorIterator& iter) {
     // Instead of implementing is_nan and warp_shfl_down
     // we can convert halves to float and do all the operations in float
     argmax_kernel_cuda_impl<at::Half, float>(iter);
+  } else if (iter.dtype(1) == kBFloat16) {
+    // Instead of implementing is_nan and warp_shfl_down
+    // we can convert halves to float and do all the operations in float
```
Comment (mruberry) [inline aten/src/ATen/native/cuda/ReduceMinMaxKernel.cu:79]: This would be "convert bfloat16 to float"

**Comment by mruberry** [comment]
Hey @imaginary-person, thanks for wanted to add more dtype support!

This change looks good, but I think we should scope it down because the trickiest part of this change will be test. And we should go through the end-to-end process of making a change, validating it, and landing it, before we make a lot more changes.

So let's pick one operation or two related operations, like torch.amin and torch.amax, to start with.

For the operation(s) you pick, implement an OpInfo in the op_db here:

https://github.com/pytorch/pytorch/blob/0e1c5cb35475935073569a2087f5fb5b7a8e673c/torch/testing/_internal/common_methods_invocations.py#L1122

One thing about implementing an OpInfo is that you need to provide sample inputs so that the test suite knows how to call the function. In the case of amin and amax there are already legacy sample input entries here:

https://github.com/pytorch/pytorch/blob/0e1c5cb35475935073569a2087f5fb5b7a8e673c/torch/testing/_internal/common_methods_invocations.py#L2459

If you take a look at other OpInfo PRs, like https://github.com/pytorch/pytorch/pull/51167/files, which creates an OpInfo for clamp, you can see that those legacy entries are removed and used as a basis for a new sample input function.

OpInfos are really helpful metadata that let the test suite generate a variety of tests, but they don't tell the test suite whether the function is working as expected or not (this is not entirely true, but for amin and amax we can assume it's the case). So there are also tests for amin and amax functionality in test_reductions.py:

https://github.com/pytorch/pytorch/blob/0e1c5cb35475935073569a2087f5fb5b7a8e673c/test/test_reductions.py#L584

These tests use the `@dtypes` decorator to test amin and amax in a variety of dtypes. You just need to add your dtypes in the appropriate decorators to get the test to work with them, too.

Does that make sense? Let me know, and ping me if you have additional questions.

I've also requested a review from @zasdfgbnm, who's been doing a lot of work improving CUDA operators' dtype support.

**Comment by imaginary-person** [comment]
Hello @mruberry, thanks a lot for the review, and for your helpful comments!

Owing to a comment in `sample_inputs_slogdet()`, I was able to figure out how to create sample inputs for `OpInfo`s,
and had been playing around with them locally today evening. I've already written `OpInfo`s for `min` & `max`, so maybe I can try adding both of them alongside `amin` & `amax`, if that's alright.

Also, as per your advice, I'll only add a few operations, so that they can be added incrementally,
and will check out `test_reductions.py`.

File: aten/src/ATen/native/cuda/ReduceMinMaxKernel.cu (hunk 1 above)
Comment (imaginary-person) [reply aten/src/ATen/native/cuda/ReduceMinMaxKernel.cu:79]: Thanks for the catch! I'll edit the comment when I'll push next time.

**Comment by mruberry** [comment]
> Hello @mruberry, thanks a lot for the review, and for your helpful comments!
> 
> Owing to a comment in `sample_inputs_slogdet()`, I was able to figure out how to create sample inputs for `OpInfo`s,
> and had been playing around with them locally today evening. I've already written `OpInfo`s for `min` & `max`, so maybe I can try adding both of them alongside `amin` & `amax`, if that's alright.
> 
> Also, as per your advice, I'll only add a few operations, so that they can be added incrementally,
> and will check out `test_reductions.py`.

Doing min and max is also OK if you like. You'll need to find the current min and max tests, and lookout for the legacy sample inputs in method_tests() for min and max, too, just like with amax and amin.

**Comment by imaginary-person** [comment]
Thanks again for your advice, @mruberry! I'll only test `min()` & `max()` for now. I used sample inputs from `method_tests()`.
I'll make the necessary changes & push the code after rebasing.

File: aten/src/ATen/native/cuda/ReduceMinMaxKernel.cu [hunk 2]
```diff
@@ -86,6 +90,10 @@ void argmin_kernel_cuda(TensorIterator& iter) {
     // Instead of implementing is_nan and warp_shfl_down
     // we can convert halves to float and do all the operations in float
     argmin_kernel_cuda_impl<at::Half, float>(iter);
+  } else  if (iter.dtype(1) == kBFloat16) {
+    // Instead of implementing is_nan and warp_shfl_down
+    // we can convert halves to float and do all the operations in float
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cuda/ReduceMinMaxKernel.cu:95]: Same here
```suggestion
    // we can convert bfloat16 to float and do all the operations in float
```

**Comment by imaginary-person** [comment]
Will push the following changes post-rebasing by 9 AM PST tomorrow:
1. Only enabling few ops (`min`, `max`).
2.` OpInfo`s for `min`, `max`
3.  enabling `test_min` & `test_max` in `test_reductions.py` for `float16` & `bfloat16`

Just updating status due to inactivity from my end.

**Comment by imaginary-person** [comment]
Hello @mruberry,

Two types of JIT-related failures were encountered during the `OpInfo` testing of `min` & `max`. Whenever you can, please advise if I should post issues for them, and please advise if I should use `SkipInfo`s for them for the time being. I didn't ping Xiang in this comment, as I think these failures aren't because of CUDA, but I'd ping him if you'd recommend.

1. **For dtype float32**, If `assert_autodiffed` is `True` in the `OpInfo`s of min & max, `test_variant_consistency_jit_min_cpu_float32`, `test_variant_consistency_jit_min_cuda_float32`, and their `max` counterparts  fail with the same cause for both CPU & CUDA:
```
AssertionError: False is not true : 
Failure in testing nodes' autodifferentiation. One or more nodes were expected to be autodiffed, but were not found in specified fusible/nonfusible DifferentiableGraph groups. 
Specifically:
  ['aten::min'] were not in one of the DifferentiableGraphs when they were expected to be. Did you intend for these nodes to be autodiffed? If not, remove them from the list of nonfusible nodes.
```
However, if `assert_autodiffed` is `False` (by default), these tests still fail in a catch-22 manner:
```
Failure in testing nodes' autodifferentiation. One or more nodes were not expected to be autodiffed but were found in a DifferentiableGraph or in a FusionGroup of a DifferentiableGraph. Did you intend for these nodes to be autodiffed? If so, change this test to expect autodifferentiation. 
Specifically:
  ['aten::min'] were not expected to be in one of the DifferentiableGraphs but were.
```


2. **For dtype bfloat16**, `test_variant_consistency_jit_min_cpu_bfloat16` and `test_variant_consistency_jit_max_cpu_bfloat16` fail with the same cause, although their CUDA variants pass.
Failure cause:
```
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 295, in instantiated_test
    raise rte
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 290, in instantiated_test
    result = test_fn(self, *args)
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 268, in test_wrapper
    return test(*args, **kwargs)
  File "test_ops.py", line 303, in test_variant_consistency_jit
    no_grad=not test_backward)
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_jit.py", line 77, in check_against_reference
    allow_unused=allow_unused)
  File "/opt/conda/lib/python3.6/site-packages/torch/autograd/__init__.py", line 225, in grad
    inputs, allow_unused, accumulate_grad=False)
RuntimeError: "method_name" not implemented for 'BFloat16'
```

**Comment by imaginary-person** [comment]
`test_reductions.py` uses numpy to verify functionality. However, numpy doesn't currently support `bfloat16`, so the tests in `test_reductions.py` can't test for the `bfloat16` dtype.

File: aten/src/ATen/native/cpu/ReduceAllOpsKernel.cpp [hunk 3]
```diff
@@ -74,7 +74,7 @@ static void min_all_kernel_impl(Tensor& result, const Tensor& input) {
     reduce_all_impl<int64_t>(result, input, upper_bound<int64_t>(),
       [=](int64_t a, int64_t b) -> int64_t { return min_impl(a, b); });
   } else {
-    AT_DISPATCH_ALL_TYPES(input.scalar_type(), "min_all", [&] {
+    AT_DISPATCH_ALL_TYPES_AND2(kHalf, kBFloat16, input.scalar_type(), "min_all", [&] {
```
Comment (imaginary-person) [inline aten/src/ATen/native/cpu/ReduceAllOpsKernel.cpp:77]: bool is supported by CUDA for its corresponding op, but not here for CPU. Is that intentional?

File: aten/src/ATen/native/cpu/ReduceAllOpsKernel.cpp [hunk 4]
```diff
@@ -99,7 +99,7 @@ static void max_all_kernel_impl(Tensor& result, const Tensor& input) {
     reduce_all_impl<int64_t>(result, input, lower_bound<int64_t>(),
       [=](int64_t a, int64_t b) -> int64_t { return max_impl(a, b); });
   } else {
-    AT_DISPATCH_ALL_TYPES(input.scalar_type(), "max_all", [&] {
+    AT_DISPATCH_ALL_TYPES_AND2(kHalf, kBFloat16, input.scalar_type(), "max_all", [&] {
```
Comment (imaginary-person) [inline aten/src/ATen/native/cpu/ReduceAllOpsKernel.cpp:102]: bool is supported by CUDA for its corresponding op, but not here for CPU. Is that intentional?

File: torch/testing/_internal/common_methods_invocations.py [hunk 5]
```diff
@@ -702,6 +697,74 @@ def sample_inputs_index_fill(op_info, device, dtype, requires_grad):
             samples.append(SampleInput((tensor, d, idx_nonctg, fill_val)))
     return samples
 
+def sample_inputs_max_min(op_info, device, dtype, requires_grad):
```
Comment (imaginary-person) [inline torch/testing/_internal/common_methods_invocations.py:700]: I'll make this one more concise

**Comment by mruberry** [comment]
Hey @imaginary-person, great questions! Sorry I missed this notification. If I don't respond to you in two working days please just ping me again. How's the PR going, by the way?

> 1. **For dtype float32**, If `assert_autodiffed` is `True` in the `OpInfo`s of min & max, `test_variant_consistency_jit_min_cpu_float32`, `test_variant_consistency_jit_min_cuda_float32`, and their `max` counterparts  fail with the same cause for both CPU & CUDA:
> 
> ```
> AssertionError: False is not true : 
> Failure in testing nodes' autodifferentiation. One or more nodes were expected to be autodiffed, but were not found in specified fusible/nonfusible DifferentiableGraph groups. 
> Specifically:
>   ['aten::min'] were not in one of the DifferentiableGraphs when they were expected to be. Did you intend for these nodes to be autodiffed? If not, remove them from the list of nonfusible nodes.
> ```
> 
> However, if `assert_autodiffed` is `False` (by default), these tests still fail in a catch-22 manner:
> 
> ```
> Failure in testing nodes' autodifferentiation. One or more nodes were not expected to be autodiffed but were found in a DifferentiableGraph or in a FusionGroup of a DifferentiableGraph. Did you intend for these nodes to be autodiffed? If so, change this test to expect autodifferentiation. 
> Specifically:
>   ['aten::min'] were not expected to be in one of the DifferentiableGraphs but were.
> ```

I think the issue you're hitting is that max is really 3 functions in one (because it's terrible). It's a reduction, a reduction that returns values and indices, and binary op. (amax and amin, btw, don't suffer from this issue.) Because max is really 3 operations you probably need to separate them into 3 OpInfos with SampleInputs that correspond to each version. If we look at the legacy max entires in method_tests(), for example:

https://github.com/pytorch/pytorch/blob/0620c96fd6a140e68c49d68ed14721b1ee108ecc/torch/testing/_internal/common_methods_invocations.py#L2630

We can see (well, if we spend the time to read these very hard to read entries) that some are autodiffed and some are not. It looks like the reductions aren't autodiffed but the binary version of the operation is. I think that's why both settings are giving you errors currently.

The 3 OpInfos are basically copies of each other with different sample input functions. Creating multiple OpInfos for a single op might seem weird, but it's an existing pattern. See the OpInfos for div for an example:

https://github.com/pytorch/pytorch/blob/0620c96fd6a140e68c49d68ed14721b1ee108ecc/torch/testing/_internal/common_methods_invocations.py#L1485

Does that make sense? Let me know, and let me know how I can help more with this PR. It's really exciting to have you hammering on our test architecture!

**Comment by imaginary-person** [comment]
Hello @mruberry, thanks a lot for your help!

I figured you were busy as I had pinged you twice on the other PR, so I didn't ping you again on this one, but would do so in the future, as per your advice.

That detailed explanation was very helpful. I'm sorry I missed that nuance!
I'll add 3 separate `OpInfo`s for `max` & `min`, and would push changes after testing locally, so that the PR can actually be ready for review by tomorrow.

**Comment by mruberry** [comment]
> Hello @mruberry, thanks a lot for your help!
> 
> I figured you were busy as I had pinged you twice on the other PR, so I didn't ping you again on this one, but would do so in the future, as per your advice.

Yes, sorry, we're finalizing our PyTorch 1.8 branch, so it's very hectic! Thank you for being so considerate, but I am here to help you so make sure I'm doing that!
 
> That detailed explanation was very helpful. I'm sorry I missed that nuance!
> I'll add 3 separate `OpInfo`s for `max` & `min`, and would push changes after testing locally, so that the PR can actually be ready for review by tomorrow.

Awesome! Looking forward to it.

**Comment by imaginary-person** [comment]
Hello @mruberry, thank you for your help! Based on your advice, I added 3 different `OpInfo`s for `min` & `max`.
Whenever you can, please help me with an issue related to `pybind11`.

**For dtype bfloat16**, `test_variant_consistency_jit` fails on CPU but passes on CUDA when a dimension argument is provided for both `min` & `max`.
Failure cause:
```
RuntimeError: "method_name" not implemented for 'BFloat16'
```

With GDB, I discovered that the exception is thrown by `pytorch/third_party/pybind11/include/pybind11/pybind11.h`.

**Comment by mruberry** [comment]
> Hello @mruberry, thank you for your help! Based on your advice, I added 3 different `OpInfo`s for `min` & `max`.
> Whenever you can, please help out me with an issue related to a JIT test.
> 
> **For dtype bfloat16**, `test_variant_consistency_jit` fails on CPU but passes on CUDA when a dimension argument is provided for both `min` & `max`. 

Unfortunately that test can be flaky. You can skip it on CPU/bfloat16 with a SkipInfo, like this one:

```
SkipInfo(
  'TestCommon', 'test_variant_consistency_jit',
  device_type='cpu', dtypes=[torch.bfloat16]),
```

File: aten/src/ATen/native/cpu/ReduceAllOpsKernel.cpp (hunk 3 above)
Comment (imaginary-person) [reply aten/src/ATen/native/cpu/ReduceAllOpsKernel.cpp:77]: Oops, bool has been handled above.

File: aten/src/ATen/native/cpu/ReduceAllOpsKernel.cpp (hunk 4 above)
Comment (imaginary-person) [reply aten/src/ATen/native/cpu/ReduceAllOpsKernel.cpp:102]: Oops, bool has been handled above.

File: test/test_reductions.py [hunk 6]
```diff
@@ -566,13 +566,13 @@ def get_values(x):
                         self.assertEqual(i, index)
                 self.assertEqual(torchfn(x), nan)
 
-    @dtypesIfCPU(torch.float, torch.double, torch.long, torch.bool)
+    @dtypesIfCPU(torch.float, torch.double, torch.long, torch.bool, torch.half)
```
Comment (imaginary-person) [inline test/test_reductions.py:569]: These tests can't be enabled for `bfloat16`, as they use `numpy` as reference, but it doesn't currently support `bfloat16`.

As a sidenote, I noticed that these tests haven't been enabled for `double` on CUDA, although they pass on CUDA.

**Comment by imaginary-person** [comment]
Thanks for your advice, @mruberry!
I had been skipping the test in the PR with a `SkipInfo`, but added a reference in comments today.
If other `max`, `min` related operations are to be added in separate PRs, then this particular PR is ready for review.

The current CI failures are due to errors in CI, so I'll trigger CI again.
```
ls: cannot access '/c/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v10.1/bin/nvcc.exe': No such file or directory
+ echo 'CUDA installation failed'
CUDA installation failed 
```

**Comment by imaginary-person** [comment]
Hello @anjali411, I just wanted to clarify that the pending checks have passed but their status wasn't updated, as GitHub rate-limited CircleCI, so the latter wasn't allowed to post an update. Thank you!

**Review by anjali411 — APPROVED** [review]
LGTM! thank you @imaginary-person for working on this

**Comment by imaginary-person** [comment]
> LGTM! thank you @imaginary-person for working on this

Thanks to @mruberry for his guidance, and thanks to you for the review!
BTW, tomorrow, I'll submit separate PRs for other min/max related ops. Thanks!

**Comment by anjali411** [comment]
@imaginary-person looks like there's a merge conflict. could you please rebase?

**Comment by imaginary-person** [comment]
Hello @anjali411, I resolved the merge-conflict. I'll submit the other PRs in a couple of hours. Thanks!

**Comment by imaginary-person** [comment]
The CI failures are unrelated to the change. They're failing for other PRs as well, such as https://app.circleci.com/pipelines/github/pytorch/pytorch?branch=pull%2F52462.

**Pull request closed without merging**